### PR TITLE
Add cume_dist Presto window function

### PR DESCRIPTION
**Pull request by pramodsatya**

### CODE REVIEW

File: velox/functions/prestosql/window/tests/CumeDistTest.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/prestosql/window/tests/WindowTestBase.h"
+
+using namespace facebook::velox::exec::test;
+
+namespace facebook::velox::window::test {
+
+namespace {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/CumeDistTest.cpp:22]: You could combine these tests with RankTest as this function too should always go along all of them.

File: velox/functions/prestosql/window/CumeDist.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  void apply(
+      const BufferPtr& peerGroupStarts,
+      const BufferPtr& peerGroupEnds,
+      const BufferPtr& /*frameStarts*/,
+      const BufferPtr& /*frameEnds*/,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    int numRows = peerGroupStarts->size() / sizeof(vector_size_t);
+    auto* rawValues = result->asFlatVector<double>()->mutableRawValues();
+    auto* peerGroupStartsValue = peerGroupStarts->as<vector_size_t>();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/CumeDist.cpp:44]: Nit -- Rename variables to peerGroupStartsVector and peer GroupEndsVector.

File: velox/functions/prestosql/window/CumeDist.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    auto* peerGroupEndsValue = peerGroupEnds->as<vector_size_t>();
+
+    for (int i = 0; i < numRows; i++) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/CumeDist.cpp:47]: Nit -- Maybe peerStart or peerGroupStart might be more consistent name.

**Review by aditi-pandit — COMMENTED** [review]
Thanks Pramod for this function.

File: velox/functions/prestosql/window/tests/CumeDistTest.cpp (hunk 1 above)
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/CumeDistTest.cpp:22]: Thanks for the suggestion, that is what I wanted to do initially actually, I will combine these!

**Comment by aditi-pandit** [comment]
Please rebase after https://github.com/facebookincubator/velox/pull/2700 lands

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 4]
```diff
@@ -36,6 +36,7 @@ TEST_F(RankTest, basic) {
   testTwoColumnInput({vectors}, "rank");
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/RankTest.cpp]: Thanks for the suggestion, changed accordingly!

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 5]
```diff
@@ -21,7 +21,28 @@ namespace facebook::velox::window::test {
...
+
+  void testWindowFunction(
+      const std::vector<RowVectorPtr>& vectors,
+      const std::vector<std::string>& overClauses) {
+    for (auto rankFunction : rankFunctions_) {
+      WindowTestBase::testWindowFunction(vectors, rankFunction, overClauses);
+    }
+  }
+
+  std::vector<std::string> rankFunctions_;
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/RankTest.cpp:44]: Can you initialize here itself and make it a const vector ?

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 6]
```diff
@@ -21,7 +21,28 @@ namespace facebook::velox::window::test {
...
 
-class RankTest : public WindowTestBase {};
+class RankTest : public WindowTestBase {
+ protected:
+  RankTest() {
+    rankFunctions_ = {"rank", "dense_rank", "percent_rank", "cume_dist"};
+  }
+
+  void testTwoColumnInput(const RowVectorPtr vectors) {
+    for (auto rankFunction : rankFunctions_) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/RankTest.cpp:31]: Make this const auto&

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 7]
```diff
@@ -21,7 +21,28 @@ namespace facebook::velox::window::test {
...
+      WindowTestBase::testTwoColumnInput({vectors}, rankFunction);
+    }
+  }
+
+  void testWindowFunction(
+      const std::vector<RowVectorPtr>& vectors,
+      const std::vector<std::string>& overClauses) {
+    for (auto rankFunction : rankFunctions_) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/RankTest.cpp:39]: Make this const auto& also

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 8]
```diff
@@ -21,7 +21,28 @@ namespace facebook::velox::window::test {
 
 namespace {
 
-class RankTest : public WindowTestBase {};
+class RankTest : public WindowTestBase {
+ protected:
+  void testTwoColumnInput(const RowVectorPtr vectors) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/tests/RankTest.cpp:26]: const &

File: velox/functions/prestosql/window/tests/RankTest.cpp [hunk 9]
```diff
@@ -21,7 +21,28 @@ namespace facebook::velox::window::test {
...
+
+  void testWindowFunction(
+      const std::vector<RowVectorPtr>& vectors,
+      const std::vector<std::string>& overClauses) {
+    for (const auto& rankFunction : rankFunctions_) {
+      WindowTestBase::testWindowFunction(vectors, rankFunction, overClauses);
+    }
+  }
+
+  const std::vector<std::string> rankFunctions_ = {
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/tests/RankTest.cpp:40]: Consider turning this test into parameterized test using function name as a parameter.

https://github.com/google/googletest/blob/main/docs/advanced.md#value-parameterized-tests

**Review by mbasmanova — COMMENTED** [review]
@aditi-pandit Looks good to me % a couple of small comments.

**Comment by aditi-pandit** [comment]
> @aditi-pandit Looks good to me % a couple of small comments.

@pramodsatya is the author of this PR. Sorry for the confusion. I only started your review after mine.

File: velox/functions/prestosql/window/tests/RankTest.cpp (hunk 9 above)
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/RankTest.cpp]: Thanks for the suggestion, modified accordingly!

**Review by mbasmanova — APPROVED** [review]
@pramodsatya Thanks. I think you need to wait for #2702 to land, then rebase. Ping me after that so I can merge.

**Comment by pramodsatya** [comment]
@mbasmanova squashed commits and rebased on top of #2702 changes, thank you!

**Comment by aditi-pandit** [comment]
https://github.com/facebookincubator/velox/issues/685

**Pull request closed without merging**